### ISSUE TITLE
Can't use jit_model_args for multi chains

### ISSUE THREAD

**Issue by fehiepsi** [pyro-ppl/numpyro#691]
@vedal observed it in #514: 

> set jit_model_args=True. However, I tried it on the Time Series Forecasting example, and it gives shape errors.

**Comment by vedal** [comment]
Thank you so much @fehiepsi for amazing responsiveness and lightning-fast PR. This seems to partially solve my problem, but I unfortunately hit another error in the mentioned tutorial (at least on my side) when running with `jit_model_args=True`:
https://github.com/google/jax/blob/master/jax/numpy/lax_numpy.py#L3774

The error seems to be due to the parameter `seasonality` in function `sgt`. When I replace it with setting a constant inside `sgt`, I manage to perform the jit-compile.

Current setup:
numpyro: git+https://github.com/fehiepsi/numpyro.git@jitarg
`jax==0.1.73`
`jaxlib==0.1.51`

**Comment by fehiepsi** [comment]
Thanks @vedal! I see that jitting index/length arguments might not be permitted. Beside your solution, there are other options such as:
+ set `seasonality=38` by default and don't feed it in `mcmc.run`
+ equivalently, set `model=partial(sgt, seasonality=38)` for NUTS.

Because the tutorial does not use `jit_model_args`, I think it is better to leave it as-is. :) Do you think `jit_model_args` helps for running multi-chains?

**Comment by vedal** [comment]
@fehiepsi I like the `model=partial(...` solution, thanks for the tip!
> Do you think jit_model_args helps for running multi-chains?
The tutorial-example runs fast for me, but I'm not sure if my setup is optimal. At the moment, I use a 24-core machine with 
4x RTX2080 GPU. Using the gpus (with `numpyro.set_host_device_count(4)`) and `num_chains=4`, I get nowhere the speed of using `cpu` and device count 24 and `num_chains=24`. The GPUs take forever. Not sure if I did something wrong, and if `numpyro.infer.mcmc` is supposed to be faster on GPU than CPU...  do you have any insight on this?

**Comment by fehiepsi** [comment]
GPU is slow for small-tensor operators. That might be the main reason I guess. If your model involves large tensor ops, then GPU will be faster. (Most of the operators in `sgt` model are scalar, so CPU is much faster there)

> not sure if my setup is optimal

You can simply set `jit_model_args=True/False` instead of using GPU. I don't observe the difference between those options.

**Comment by vedal** [comment]
Thanks alot for your feedback, explanations and PR @fehiepsi and @neerajprad

**Comment by ecotner** [comment]
Hello, I'm running into this problem on version `0.7.2` of numpyro, which is strange, since it seems from the above conversation that the fix had been merged. Here is a summary of my code (all arguments to the model are JAX arrays and the `N*` are integers):

```python
def model(sid, X, y=None):
    # sample hyperpriors
    with numpyro.plate("features", Nf):
        μ = numpyro.sample("mu", Normal(0.25, 1))
        σ = numpyro.sample("sigma", HalfNormal(1.5))
        # sample coefficients
        with numpyro.plate("drivers", Nd):
            β = numpyro.sample("beta", Normal(μ, σ))
    # calculate probabilities
    p = jax.nn.sigmoid((β[sid] * X).sum(axis=1))
    # sample target variable
    with numpyro.plate("data", N):
        yhat = numpyro.sample("y", BernoulliProbs(p), obs=y)
    return yhat

kernel = NUTS(model, init_strategy=numpyro.infer.init_to_median)
mcmc = MCMC(kernel, num_warmup=10, num_samples=10, num_chains=2, jit_model_args=True)
rng_key = jax.random.PRNGKey(0)
mcmc.run(rng_key, sid, X, y)  # run first time
mcmc._warmup_state = mcmc._last_state
mcmc.run(rng_key, sid, X, y)  # run second time
```
When I use `mcmc.run` the second time, it does skip the warmup phase, but it still recompiles!

**Comment by fehiepsi** [comment]
Hi @ecotner, this issue has been fixed to make your code run. I think "speed"/"caching" is another issue. For caching with parallel chain, please see the issue https://github.com/pyro-ppl/numpyro/issues/1134 .
